client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,338 @@
+        Ok(data) => Some(data),
+        Err(e) => {
+            eprintln!("Failed to load SMBIOS data: {}.", e);
+            None
+        }
+    };
+    Ok(smbios_data)
+}
+
+pub fn collect_bios_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<devices::Bios, Box<dyn std::error::Error>> {
+    let bios_info = &smbios_data.collect::<smbioslib::SMBiosInformation>()[0];
+
+    let release_date_str = bios_info.release_date().to_string();
+    let release_date_format = format_description!("[month]/[day]/[year]");
+    let release_date_parsed = Date::parse(&release_date_str, &release_date_format)?;
+    let output_format = format_description!("[year]-[month]-[day]");
+    let release_date = release_date_parsed.format(&output_format)?;
+
+    let bios = devices::Bios {
+        firmware_revision: match (
+            bios_info.system_bios_major_release(),
+            bios_info.system_bios_minor_release(),
+        ) {
+            (Some(major), Some(minor)) => Some(format!("{}.{}", major, minor)),
+            _ => None,
+        },
+        release_date: Some(release_date),
+        revision: Some(bios_info.version().to_string()),
+        vendor: bios_info.vendor().to_string(),
+        version: bios_info.version().to_string(),
+    };
+    Ok(bios)
+}
+
+/// Retrieve CPU information from SMBios
+pub fn collect_processor_info_smbios(
+    smbios_data: &smbioslib::SMBiosData,
+    max_cpu_frequency_filepath: Option<&'static str>,
+) -> Result<devices::Processor, Box<dyn std::error::Error>> {
+    let processor_info = &smbios_data.collect::<smbioslib::SMBiosProcessorInformation>()[0];
+    let cpu_id = super::cpuid::get_cpuid(processor_info)?;
+    let processor = devices::Processor {
+        codename: super::cpuid::convert_cpu_codename(&cpu_id)?,
+        frequency: super::cpuinfo::read_max_cpu_frequency(max_cpu_frequency_filepath)?,
+        manufacturer: processor_info.processor_manufacturer().to_string(),
+        version: processor_info.processor_version().to_string(),
+    };
+    Ok(processor)
+}
+
+/// Retrieve CPU information from /proc/cpuinfo
+pub fn collect_processor_info_cpuinfo(
+    cpuinfo_filepath: Option<&'static str>,
+    max_cpu_frequency_filepath: Option<&'static str>,
+) -> Result<devices::Processor, Box<dyn std::error::Error>> {
+    let cpu_info = super::cpuinfo::parse_cpuinfo(cpuinfo_filepath)?;
+    let processor = devices::Processor {
+        codename: String::new(),
+        frequency: super::cpuinfo::read_max_cpu_frequency(max_cpu_frequency_filepath)?,
+        manufacturer: cpu_info.cpu_type,
+        version: cpu_info.model,
+    };
+    Ok(processor)
+}
+
+pub fn collect_chassis_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<crate::models::devices::Chassis, Box<dyn std::error::Error>> {
+    let chassis_info = &smbios_data.collect::<smbioslib::SMBiosSystemChassisInformation>()[0];
+
+    let chassis_type = chassis_info.chassis_type().ok_or("")?.to_string();
```

Comment:
If this string is being used as an error message it should probably explain what the error is :slightly_smiling_face:

Reply:
Fixed in 9f14753